pkg/driver/controller.go

```diff
@@ -109,93 +98,85 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	}
 
 	volumeParams := req.GetParameters()
-	volumeType := volumeParams[volumeParamsVolumeType]
 
-	var storageCapacity *int64
+	err := didUserDefineDriverProvidedParameters(volumeParams)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	var tags []map[string]string
+	_, ok := volumeParams[volumeParamsTags]
+	if ok {
+		err = util.ConvertJsonStringToObject(volumeParams[volumeParamsTags], &tags)
+
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, ErrIncorrectlyFormatted, volumeParamsTags, err)
+		}
+	}
+	tags = append(tags, map[string]string{"Key": AwsFsxOpenZfsDriverTagKey, "Value": "true"})
 
+	var deleteParams = make(map[string]any)
+	for key, value := range volumeParams {
+		if strings.HasSuffix(key, "OnDeletion") {
+			deleteParameter := strings.TrimSuffix(key, "OnDeletion")
+			deleteParams[deleteParameter] = value
+			if strings.ContainsAny(value, "[],\"") {
+				value = util.EncodeDeletionTag(value)
+			}
+			tags = append(tags, map[string]string{"Key": key, "Value": value})
+			delete(volumeParams, key)
+		}
+	}
+
+	tagsString, err := util.ConvertObjectToJsonString(tags)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, ErrIncorrectlyFormatted, "Internal Tags", err)
+	}
+
+	volumeParams[volumeParamsClientRequestToken] = strconv.Quote(volName)
+	volumeParams[volumeParamsTags] = tagsString
+
+	for key, _ := range volumeParams {
+		if strings.HasPrefix(key, reservedVolumeParamsPrefix) {
+			delete(volumeParams, key)
+		}
+	}
+
+	var storageCapacity int64
 	if req.GetCapacityRange() != nil {
-		convertedCapacity := util.BytesToGiB(req.GetCapacityRange().GetRequiredBytes())
-		storageCapacity = &convertedCapacity
+		storageCapacity = util.BytesToGiB(req.GetCapacityRange().GetRequiredBytes())
 	}
 
-	if volumeType == fsVolumeType {
+	resourceType := volumeParams[volumeParamsType]
+	delete(volumeParams, volumeParamsType)
+	if resourceType == fsVolumeType {
 		if req.GetVolumeContentSource() != nil {
-			return nil, status.Error(codes.InvalidArgument, "Snapshots are not available at the filesystem level. Set volumeType to volume.")
+			return nil, status.Error(codes.InvalidArgument, "Snapshots are not available at the filesystem level. Set resourceType to volume.")
 		}
 
-		if _, ok := volumeParams[volumeParamsSkipFinalBackupOnDeletion]; !ok {
-			return nil, status.Errorf(codes.InvalidArgument, "invalid parameter of %s: %s", volumeParamsSkipFinalBackupOnDeletion, "field required")
+		if _, ok := deleteParams[volumeParamsSkipFinalBackup]; !ok {
+			return nil, status.Errorf(codes.InvalidArgument, ErrIncorrectlyFormatted, volumeParamsSkipFinalBackup, "field is required")
 		}
 
-		fsOptions := cloud.FileSystemOptions{
-			StorageCapacity: storageCapacity,
+		deleteParams["FileSystemId"] = "fs-1234567890abc"
```

Comment:
Is this left in by mistake? Why do we need this placeholder value

Reply:
It's intentional. Since it's in create we don't have a FileSystemId yet. Calling validate on DeleteFileSystemInput throws an error since FileSystemId is a required field. Therefore we want to populate it with a filler. Then validate should only error if the user didn't input a correct field.